pkg/transform/transform.go

```diff
@@ -0,0 +1,158 @@
+
+type Cluster struct {
+	Master Master
+}
+
+type Master struct {
+	OAuth   oauth.OAuthCRD
+	Secrets []secrets.Secret
+}
+
+type Manifest struct {
+	Name string
+	CRD  []byte
+}
+
+type Config struct {
+	MasterConfigFile     string
+	NodeConfigFile       string
+	RegistriesConfigFile string
+	OutputDir            string
+	Hostname             string
+}
+
+type ManifestTransformOutput struct {
+	//Config    Config
+	Manifests []Manifest
+}
+
+type TransformRunner struct {
+	Config string
+}
+
+// GetFile allows to mock file retrieval
+var GetFile = io.GetFile
+
+func Start() {
+	config := Config{}
+	config.OutputDir = env.Config().GetString("OutputDir")
+	config.Hostname = env.Config().GetString("Source")
+	config.MasterConfigFile = MasterConfigFile
+	config.RegistriesConfigFile = RegistriesConfigFile
+	transformRunner := NewTransformRunner(config)
+
+	if err := transformRunner.Transform([]Transform{
+		OAuthTransform{
+			Config: &config,
+		},
+		SDNTransform{
+			Config: &config,
+		},
+		RegistriesTransform{
+			Config: &config,
+		},
+	}); err != nil {
+		logrus.WithError(err).Fatalf("%s", err.Error())
+	}
+}
+
+// DumpManifests creates OCDs files
+func DumpManifests(manifests []Manifest) {
+	for _, manifest := range manifests {
+		maniftestfile := filepath.Join(env.Config().GetString("OutputDir"), "manifests", manifest.Name)
+		os.MkdirAll(path.Dir(maniftestfile), 0755)
+		err := ioutil.WriteFile(maniftestfile, manifest.CRD, 0644)
+		logrus.Printf("CRD:Added: %s", maniftestfile)
+		if err != nil {
+			logrus.Panic(err)
+		}
+	}
+}
+
+func (config *Config) Fetch(path string) []byte {
+	dst := filepath.Join(config.OutputDir, config.Hostname, path)
+	f := GetFile(config.Hostname, path, dst)
+	logrus.Printf("File:Loaded: %s", dst)
+
+	return f
+}
+
+type Extraction interface {
+	Transform() (TransformOutput, error)
+	Validate() error
+}
+
+type Transform interface {
+	Extract() Extraction
+}
+
+type TransformOutput interface {
+	Flush() error
+}
+
+func (m ManifestTransformOutput) Flush() error {
+	logrus.Info("Writing file data:")
+	DumpManifests(m.Manifests)
+	return nil
+}
+
+func (r TransformRunner) Transform(transforms []Transform) error {
+	logrus.Info("TransformRunner::Transform")
+
+	// For each transform, extract the data, validate it, and run the transform.
+	// Handle any errors, and finally flush the output to it's desired destination
+	// NOTE: This should be parallelized with channels unless the transforms have
+	// some dependency on the outputs of others
+	for _, transform := range transforms {
+		extraction := transform.Extract()
+
+		if err := extraction.Validate(); err != nil {
+			HandleError(err)
+		}
+
+		output, err := extraction.Transform()
+		if err != nil {
+			HandleError(err)
+		}
+
+		if err := output.Flush(); err != nil {
+			HandleError(err)
+		}
+	}
+
+	return nil
+}
+
+func NewTransformRunner(config Config) *TransformRunner {
+	return &TransformRunner{}
+}
+
+func HandleError(err error) error {
```

Comment:
[ISSUE] I had this in the POC as the top level error handler just to illustrate typical go error handling and bubbling. We should reconsider this (I know @alexander-demichev has a PR out that is seeking to address it).

Reply:
Agree. I was hoping we could reuse @alexander-demichev's work. Unsure though if it's going to be a major rework. I can try to help with that if necessary.

Reply:
@jmontleon Feel free to do anything you think is necessary. I will rebase or rework PR with error handling.